src/HDF5.jl

```diff
@@ -1062,6 +1102,89 @@ function generic_read(obj::DatasetOrAttribute, filetype::Datatype, ::Type{T}, I.
     end
 end
 
+
+"""
+    similar(obj::DatasetOrAttribute, [::Type{T}], [I::Integer...])
+
+Return a `Array{T}` or `Matrix{UInt8}` to that can contain [part of] the dataset.
+"""
+function Base.similar(obj::DatasetOrAttribute, I::Integer...)
```

Comment:
Question, why do we need to pass the indices here?

Reply:
`Base.similar` accepts an optional size value. These are not indices. I suppose we should rename the argument.
https://docs.julialang.org/en/v1/base/arrays/#Base.similar

Reply:
I was more specifically looking at the AbstractArray interface https://docs.julialang.org/en/v1/manual/interfaces/#man-interface-array, where there are 4 optional `similar` methods that we could define:
<html>
<body>
<!--StartFragment-->

similar(A) | similar(A, eltype(A), size(A)) | Return a mutable array with the same shape and element type
-- | -- | --
similar(A, ::Type{S}) | similar(A, S, size(A)) | Return a mutable array with the same shape and the specified element type
similar(A, dims::Dims) | similar(A, eltype(A), dims) | Return a mutable array with the same element type and size dims
similar(A, ::Type{S}, dims::Dims) | Array{S}(undef, dims) | Return a mutable array with the specified element type and size

<!--EndFragment-->
</body>
</html>

Reply:
When we do implement AbstractArray in 0.17, then we do not need this signature. We could just implement the version with Dims. However, since we are not implementing AbstractArray in 0.16, we probably do need to retain a method with the above signature since we do not have a `similar(a::AbstractArray{T}, dims::Union{Integer, AbstractUnitRange}...)` to cover that case for us.

I will rename this to `Base.similar(obj::DatasetOrAttribute, dims::Dims)`. We probably sill need `Base.similar, dims::Integer...)` as well, until we actually do implement `AbstractArray`.

Reply:
@simonbyrne  did you intend to comment on the other thread?   Is there a specific part of https://github.com/JuliaIO/HDF5.jl/pull/937#discussion_r888036072 you disagree with?  
The goal is not to remove `copyto!` and replace it with `read!`. Originally, @mkitti  had `read!` as an alias for `copyto!`, which we require in order to satisfy part of the `AbstractArray` interface. My comment on `read!` was that this was both redundant with `copyto!` and unexpected since `read!` is used in base Julia for reading from an IO stream/filename, which is very different than what `copyto!` accomplishes.